snuba/optimize_tracker.py

```diff
@@ -0,0 +1,69 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Optional, Set
+
+from snuba.redis import RedisClientType
+
+OPTIMIZE_PREFIX = "snuba-optimize"
+
+
+class OptimizedPartitionTracker(ABC):
+    """
+    OptimizedPartitionTracker is an abstract class for tracking progress of an optimize job.
+    """
+
+    @abstractmethod
+    def get_completed_partitions(self) -> Optional[Set[str]]:
+        """
+        Get a list of partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def update_completed_partitions(self, part_name: str) -> None:
+        """
+        Add partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_all_partitions(self) -> None:
+        """
+        Remove all partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+
+class RedisOptimizedPartitionTracker(OptimizedPartitionTracker):
+    """
+    This class keeps track of partitions which have already been optimized by keeping state of
+    optimized partitions in redis.
+    """
+
+    def __init__(
+        self, redis_client: RedisClientType, host: str, expire_time: datetime
+    ) -> None:
+        self.__redis_client = redis_client
+        self.__bucket = f"{OPTIMIZE_PREFIX}:{host}"
```

Comment:
I think you should also add the table. Today we only have one use case, but this is not granted. We may want (or more likely be forced to because I would definitively not want) to optimize multiple tables on the same host.

Reply:
Done